src/odemis/driver/keysight.py

```diff
@@ -420,44 +461,58 @@ def _set_duty_cycle(self, dc: float) -> float:
         return act_dc
 
     def _set_power(self, p: bool) -> bool:
-        # Note: when power is off, the sync is still sent, but not waveform signal is put back to 0V.
-        self.setOutput(self._channel, p)
-
-        for c in self._tracking.keys():
-            self.setOutput(c, p)
-
-        act_p = self.getOutput(self._channel)
-        return act_p
+        if p == self.power.value:
+            # Make sure to not send the commands again if the device is already in the right state.
+            # Especially important to not increase the lock count too high, which would prevent the
+            # device from being unlocked on power off.
+            logging.debug("Not changing the power state, already %s", p)
+            return p
 
-    def _update_settings(self):
-        """
-        Update the settings of the device.
-        Called by the repeating timer.
-        """
-        try:
-            f = self.getFrequency(self._channel)  # Device always returns f > 0
-            p = 1 / f
-            if p != self.period.value:
-                self.period._set_value(p)  # Avoid the setter
-
-            # Duty cycle
-            dc = self.getDutyCycle(self._channel) / 100
-            if dc != self.dutyCycle.value:
-                self.dutyCycle._set_value(dc)
-
-            # Delay
-            act_phase = self.getPhase(self._channel)
-            d = self.period.value * act_phase / 360
-            if d != self.delay.value:
-                self.delay._set_value(d)
-
-            # Power
-            pw = self.getOutput(self._channel)
-            if pw != self.power.value:
-                self.power._set_value(pw)
-        except Exception:
-            logging.exception("Failed to update the settings")
+        if p:
+            self.setWaveform(self._channel, "SQU")
+            for c, t in self._tracking.items():
+                other_c = 3 - c
+                self.setTracking(other_c, t)
+
+            for c, lim in enumerate(self._limits):
+                self.setVoltageMin(c + 1, lim[0])
+                self.setVoltageMax(c + 1, lim[1])
+
+            # Reset all the settings, for 3 reasons:
+            # * if the VAs were changed while the power was off, they couldn't be applied, so we
+            # need to set them now
+            # * maybe the user has changed them in the meantime. So we need to put back the expected
+            # values.
+            # * it actually might be the first time the device is powered on, so we need to set them
+            # all.
+            self.dutyCycle.value = self.dutyCycle.value
```

Comment:
This is very cryptic and the setter is also very vague to me. If the power is off, we still set the value. Is this value then different than what is on the machine? If the power is on, some data alteration takes place (multiplying by 1e2), which is also under-explained in my opinion.

It would require less comments if we bookkeep a shadow value (i.e. self.dutyCycleTarget) and keep the dutyCycle VA always in sync with the device.

Reply:
You're right, at this point, when changing the power from off -> on, the VA value is still "False". So calling the setter of dutyCycle will have no effect.